src/algebra/free.lean

```diff
@@ -206,58 +360,83 @@ end lift
 
 variables {β : Type v} [has_mul β] (f : α → β)
 
+/-- From a magma homomorphism `α → β` to a semigroup homomorphism `magma.free_semigroup α → magma.free_semigroup β`. -/
+@[to_additive "From an additive magma homomorphism `α → β` to an additive semigroup homomorphism `add_magma.free_add_semigroup α → add_magma.free_add_semigroup β`."]
 def map (hf : ∀ x y, f (x * y) = f x * f y) : free_semigroup α → free_semigroup β :=
 lift (of ∘ f) (λ x y, congr_arg of $ hf x y)
 
-@[simp] lemma map_of {hf} (x) : map f hf (of x) = of (f x) := rfl
-@[simp] lemma map_mul {hf} (x y) : map f hf (x * y) = map f hf x * map f hf y :=
+@[simp, to_additive] lemma map_of {hf} (x) : map f hf (of x) = of (f x) := rfl
+@[simp, to_additive] lemma map_mul {hf} (x y) : map f hf (x * y) = map f hf x * map f hf y :=
 lift_mul _ _ _
 
 end free_semigroup
 
 end magma
 
-
+/-- Free semigroup over a given alphabet. -/
```

Comment:
Could you recall in the docstring that these do not include the empty word? I think there is an ambiguity about this in the maths litterature (in my domain, the free semigroup contains the empty word), so better be too precise.

Reply:
If A is the free semigroup on one variable (x) with the empty word and B is the free semigroup on one variable (x) without the empty word, then which function A -> B should the function {x} -> B sending x to x lift to?

Reply:
@kckennylau, I don't think @sgouezel is arguing over the "correct" definition; just that some parts of the maths literature use the phrase differently (which is indeed "incorrect" from an algebraist's or category theorist's point of view), and so we need a comment here.

Reply:
```suggestion
/-- Free semigroup over a given alphabet.
(Note: In this definition, the free semigroup does not contain the empty word.) -/
```

Reply:
That said, I'm failing to find any references, including in dynamics, that define free semigroups to contain the empty word...

Reply:
I guess the phrases "translation semigroup" and "one-parameter semigroup", which are both used widely, really mean "a monoid map from `[0,\infty)` to X" (and never intend that you exclude zero).

Reply:
A random example is https://arxiv.org/pdf/1712.00860.pdf (where on Page 2 they write "let us fix a finite word $w \ne e$ in $\Sigma$, which means implicitly that for them $e$ is an element of $\Sigma$). Or https://arxiv.org/abs/1211.6242